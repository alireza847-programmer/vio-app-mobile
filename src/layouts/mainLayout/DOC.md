# MainLayout Component

## Overview

The `MainLayout` component is a flexible layout structure designed to provide a consistent and customizable layout for various sections of your application. It includes options for a header, title, close button, and a customizable right icon.

## How to Use

1. **Import the Component**:

   ```jsx
   import MainLayout from 'path/to/MainLayout';
   ```

2. **Use the Component**:

   ```jsx
   const YourComponent = () => {
     return (
       <MainLayout
         withHeader={true} // Set to false to hide the header
         title="Your Title" // Set the title for the header
         withClose={true} // Set to false to hide the close button
         rightIcon={<YourIcon />} // JSX for the right icon
         withLogo={false} // Set to true to display a logo instead of text title
       >
         {/* Your content goes here */}
       </MainLayout>
     );
   };
   ```

## Props

- `children` (ReactNode): The content to be rendered within the layout.

- `title` (string, optional): The title to be displayed in the header. If `withLogo` is set to true, the title can also be a custom logo.

- `withClose` (boolean, optional): If true, a close button will be displayed in the header for closing the layout.

- `rightIcon` (ReactNode, optional): JSX for a customizable icon to be displayed on the right side of the header.

- `withHeader` (boolean, optional): If true, a header will be rendered at the top of the layout.

- `withLogo` (boolean, optional): If true, the title will be treated as a logo, and the default text styling will be overridden.

- `onClosePress` (function, optional): callBack function of close button

## Example

```jsx
import React from 'react';
import MainLayout from 'path/to/MainLayout';

const YourComponent = () => {
  return (
    <MainLayout
      withHeader={true}
      title="Your Title"
      withClose={true}
      rightIcon={<YourIcon />}
      withLogo={false}>
      {/* Your content goes here */}
    </MainLayout>
  );
};

export default YourComponent;
```

## Notes

- The `MainLayout` component provides a consistent structure for your application's sections, making it easy to manage and customize headers and content.
