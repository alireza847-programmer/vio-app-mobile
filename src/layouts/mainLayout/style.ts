import styled from '@emotion/native';
import VRow from 'components/uiElements/row';
import {wp} from 'utils/dimensions';

export const Container = styled.View(({theme}) => ({
  width: '100%',
}));

export const Header = styled(VRow)(({theme}) => ({
  width: wp(100),
  shadowColor: 'rgba(68, 80, 95, 0.20)',
  shadowOpacity: 1,
  shadowOffset: {
    height: 0,
    width: 0,
  },
  backgroundColor: theme.colors.backgroundColor,
  paddingTop: theme.verticalRem * 4,
  paddingBottom: theme.verticalRem * 6,
  alignItems: 'flex-start',
  elevation: 10,
}));

export const ChildrenWrapper = styled.View(({theme}) => ({
  paddingHorizontal: theme.HorizontalRem * 4,
}));
