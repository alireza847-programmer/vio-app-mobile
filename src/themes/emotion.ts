import {EmotionThemeType} from 'types/themes/emotion';
import {hp, wp} from 'utils/dimensions';

const Fonts = {
  bold: 'Inter-Bold',
  medium: 'Inter-Medium',
  semiBold: 'Inter-SemiBold',
  regular: 'Inter-Regular',
};

export const theme: EmotionThemeType = {
  colors: {
    backgroundColor: '#F2F2F2',
    textColor: {
      primary: '#2A333D',
      placeHolder: '#A3ACB8',
      button: {
        primary: '#FFFFFF',
        secondary: '#0071F3',
        error: '#D83B3B',
      },
    },
    button: {
      primary: '#0071F3',
      secondary: '#F7FBFF',
    },
    stroke: {
      button: '#DAE9FA',
      divider: '#EFF2F6',
      input: '#BED2E9',
    },
  },
  fonts: {
    bold21: {
      fontFamily: Fonts.bold,
      fontSize: 21,
    },
    medium15: {
      fontFamily: Fonts.medium,
      fontSize: 15,
    },
    semiBold16: {
      fontFamily: Fonts.semiBold,
      fontSize: 16,
    },
    semiBold18: {
      fontFamily: Fonts.semiBold,
      fontSize: 18,
    },
    regular16: {
      fontFamily: Fonts.regular,
      fontSize: 16,
    },
  },
  verticalRem: hp(0.5),
  HorizontalRem: wp(1),
  radius: wp(1),
};
