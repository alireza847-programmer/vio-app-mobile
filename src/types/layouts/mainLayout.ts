export interface MainLayoutProps {
  withHeader?: boolean;
  withLogo?: boolean;
  title?: string;
  withClose?: boolean;
  rightIcon?: JSX.Element;
}
