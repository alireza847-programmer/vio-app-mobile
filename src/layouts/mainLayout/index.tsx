import React, {PropsWithChildren} from 'react';
import {ChildrenWrapper, Container, Header} from './style';
import VText from 'components/uiElements/text';
import {MainLayoutProps} from 'types/layouts/mainLayout';
import VRow from 'components/uiElements/row';
import {CloseSvg, LogoSvg} from 'assets/svgs';
import {theme} from 'themes/emotion';

const MainLayout = (props: PropsWithChildren<MainLayoutProps>) => {
  const {
    children,
    title = '',
    withClose = true,
    rightIcon,
    withHeader,
    withLogo,
  } = props;
  const renderHeaderTitle = () => {
    if (withLogo) {
      return <LogoSvg />;
    }
    return title;
  };
  return (
    <Container>
      {withHeader && (
        <Header
          paddingHorizontalRatio={3}
          alignItems="center"
          justifyContent="space-between">
          <VRow fullWidth={false}>
            {withClose && <CloseSvg fill={theme.colors.button.primary} />}
          </VRow>
          <VText typography="semiBold16">{renderHeaderTitle()}</VText>
          <VRow fullWidth={false}>{rightIcon}</VRow>
        </Header>
      )}
      <ChildrenWrapper>{children}</ChildrenWrapper>
    </Container>
  );
};
export default MainLayout;
