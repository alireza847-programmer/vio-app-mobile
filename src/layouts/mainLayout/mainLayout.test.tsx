import React, {PropsWithChildren} from 'react';
import renderer from 'react-test-renderer';
import MainLayout from './index';
import {ThemeProvider} from '@emotion/react';
import {theme} from 'themes/emotion';
import {test, expect} from '@jest/globals';
import {MainLayoutProps} from 'types/layouts/mainLayout';

const YourIcon = () => <svg>Custom Icon</svg>;

const Component = (props: PropsWithChildren<MainLayoutProps>) => (
  <ThemeProvider theme={theme}>
    <MainLayout {...props} />
  </ThemeProvider>
);

test('MainLayout component snapshot', () => {
  const tree = renderer
    .create(<Component withHeader={true} title="Your Title" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('MainLayout component snapshot with close button and custom right icon', () => {
  const tree = renderer
    .create(
      <Component withHeader={true} withClose={true} rightIcon={<YourIcon />} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('MainLayout component snapshot with logo', () => {
  const tree = renderer
    .create(<Component withHeader={true} withLogo={true} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('MainLayout component snapshot without header', () => {
  const tree = renderer.create(<Component withHeader={false} />).toJSON();
  expect(tree).toMatchSnapshot();
});
