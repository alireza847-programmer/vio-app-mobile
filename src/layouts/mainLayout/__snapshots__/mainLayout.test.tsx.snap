// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout component snapshot 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    alignItems="center"
    justifyContent="space-between"
    paddingHorizontalRatio={3}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "paddingHorizontal": 22.5,
          "paddingVertical": 0,
          "width": "100%",
        },
        [
          {
            "alignItems": "flex-start",
            "backgroundColor": "#fff",
            "paddingBottom": 40.019999999999996,
            "paddingTop": 26.68,
            "shadowColor": "rgba(68, 80, 95, 0.20)",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "width": 750,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      fullWidth={false}
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          undefined,
        ]
      }
    >
      <SvgMock
        fill="#0071F3"
      />
    </View>
    <Text
      style={
        [
          {
            "color": "#2A333D",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
            "marginLeft": 0,
            "marginTop": 0,
            "textAlign": "center",
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
      typography="semiBold16"
    >
      Your Title
    </Text>
    <View
      fullWidth={false}
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      [
        {
          "paddingHorizontal": 30,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`MainLayout component snapshot with close button and custom right icon 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    alignItems="center"
    justifyContent="space-between"
    paddingHorizontalRatio={3}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "paddingHorizontal": 22.5,
          "paddingVertical": 0,
          "width": "100%",
        },
        [
          {
            "alignItems": "flex-start",
            "backgroundColor": "#fff",
            "paddingBottom": 40.019999999999996,
            "paddingTop": 26.68,
            "shadowColor": "rgba(68, 80, 95, 0.20)",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "width": 750,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      fullWidth={false}
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          undefined,
        ]
      }
    >
      <SvgMock
        fill="#0071F3"
      />
    </View>
    <Text
      style={
        [
          {
            "color": "#2A333D",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
            "marginLeft": 0,
            "marginTop": 0,
            "textAlign": "center",
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
      typography="semiBold16"
    />
    <View
      fullWidth={false}
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          undefined,
        ]
      }
    >
      <svg>
        Custom Icon
      </svg>
    </View>
  </View>
  <View
    style={
      [
        {
          "paddingHorizontal": 30,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`MainLayout component snapshot with logo 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    alignItems="center"
    justifyContent="space-between"
    paddingHorizontalRatio={3}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "paddingHorizontal": 22.5,
          "paddingVertical": 0,
          "width": "100%",
        },
        [
          {
            "alignItems": "flex-start",
            "backgroundColor": "#fff",
            "paddingBottom": 40.019999999999996,
            "paddingTop": 26.68,
            "shadowColor": "rgba(68, 80, 95, 0.20)",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "width": 750,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      fullWidth={false}
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          undefined,
        ]
      }
    >
      <SvgMock
        fill="#0071F3"
      />
    </View>
    <Text
      style={
        [
          {
            "color": "#2A333D",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
            "marginLeft": 0,
            "marginTop": 0,
            "textAlign": "center",
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
      typography="semiBold16"
    >
      <SvgMock />
    </Text>
    <View
      fullWidth={false}
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      [
        {
          "paddingHorizontal": 30,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`MainLayout component snapshot without header 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "paddingHorizontal": 30,
        },
        undefined,
      ]
    }
  />
</View>
`;
